File: server/core/challenge.js
/**
 * Challenge detection + resolution layer.
 *
 * The explicit "bypass JS challenges" capability. Many bot-protection systems
 * (Cloudflare, DataDome, PerimeterX) serve a JS challenge page that
 * auto-resolves after 5-15 seconds WHEN the browser fingerprint passes their
 * automation checks. The existing stealth stack (puppeteer-extra-plugin-stealth
 * + fingerprint injection) provides that detection resistance. This module
 * adds the patience: detect the challenge page, wait for self-resolution, and
 * report honestly whether it cleared.
 *
 * Three challenge classes:
 *   1. Auto-resolvable JS challenges  -> wait, re-check, extract once cleared
 *   2. Interactive challenges          -> detect, do NOT solve, fail honestly
 *   3. Hard blocks (403/captcha walls) -> detect, fail honestly
 *
 * The honest failure is load-bearing: previously a challenge page was extracted
 * as if it were content (success: true with garbage). Now unresolved challenges
 * return errorCode: "CHALLENGE_BLOCKED" so the caller knows to retry, escalate,
 * or use a solver service — instead of trusting a wrong answer.
 */

/**
 * Challenge signature. Each entry names a protection system and the signals
 * that identify its challenge page. Signals are checked in order; the first
 * match wins so the kind is reported accurately.
 *
 * @typedef {Object} ChallengeSignature
 * @property {string} kind          Machine-readable kind (cloudflare, datadome, etc.)
 * @property {string} label         Human-readable name.
 * @property {boolean} autoResolvable  Whether the JS challenge typically self-clears.
 * @property {RegExp[]} [urlPatterns]   Patterns matched against the page URL.
 * @property {string[]} [domMarkers]    CSS-like selectors checked for presence.
 * @property {RegExp[]} [textPatterns]  Patterns matched against page text.
 */

/** @type {ChallengeSignature[]} */
const CHALLENGE_SIGNATURES = [
  {
    kind: "cloudflare",
    label: "Cloudflare",
    autoResolvable: true,
    urlPatterns: [/^\/cdn-cgi\/challenge/, /\/cdn-cgi\/turnstile/],
    domMarkers: ["#cf-challenge", "#challenge-form", "#cf-please-wait", ".cf-browser-verification"],
    textPatterns: [/just a moment/i, /checking your browser/i, /cf-challenge/i, /enable javascript.*cloudflare/i],
  },
  {
    kind: "cloudflare_turnstile",
    label: "Cloudflare Turnstile",
    autoResolvable: false,
    domMarkers: ["iframe[src*='challenges.cloudflare.com']", ".cf-turnstile"],
    textPatterns: [/verify you are human/i],
  },
  {
    kind: "datadome",
    label: "DataDome",
    autoResolvable: true,
    urlPatterns: [/_dd_s\b/, /\/_dd\b/],
    domMarkers: ["iframe[src*='datadome']", "#datadome", ".dd-pp-list"],
    textPatterns: [/datadome/i, /please verify you are a person/i],
  },
  {
    kind: "perimeterx",
    label: "PerimeterX / HUMAN",
    autoResolvable: true,
    urlPatterns: [/\/_pxhl\//, /\/px\.js/],
    domMarkers: ["#px-captcha", "iframe[src*='px-captcha']", "script[src*='px-captcha']"],
    textPatterns: [/press.*hold/i, /perimeterx/i, /are you a human/i],
  },
  {
    kind: "hcaptcha",
    label: "hCaptcha",
    autoResolvable: false,
    domMarkers: ["iframe[src*='hcaptcha']", ".h-captcha", "#hcaptcha-wrapper"],
    textPatterns: [/hcaptcha/i],
  },
  {
    kind: "recaptcha",
    label: "reCAPTCHA",
    autoResolvable: false,
    domMarkers: ["iframe[src*='recaptcha']", ".g-recaptcha", "#recaptcha"],
    textPatterns: [/recaptcha/i],
  },
  {
    kind: "akamai",
    label: "Akamai Bot Manager",
    autoResolvable: true,
    urlPatterns: [/_abck\b/],
    textPatterns: [/access denied.*akamai/i, /reference.*#?\d+\.\w+.*akamai/i],
  },
  {
    kind: "generic",
    label: "Generic interstitial",
    autoResolvable: true,
    textPatterns: [
      /\bplease wait\b/i,
      /\bloading\b.*\bplease\b/i,
      /\bverifying your browser/i,
      /\bchecking.*connection/i,
    ],
  },
  {
    kind: "blocked",
    label: "Hard block",
    autoResolvable: false,
    urlPatterns: [/\/sorry\//i, /\/access[-_]?denied/i],
    textPatterns: [
      /\baccess denied\b/i,
      /\b403 forbidden\b/i,
      /\btemporarily blocked\b/i,
      /\bsecurity check\b/i,
      /\bautomated queries\b/i,
    ],
  },
];

/**
 * Inspect a page for challenge signatures. Returns the first matching kind
 * (most specific first) or null if no challenge is detected. Pure-ish over the
 * page's URL + text + DOM, so it can be re-run on a polling loop.
 *
 * @param {import("puppeteer").Page} page
 * @returns {Promise<{ kind: string, label: string, autoResolvable: boolean } | null>}
 */
export async function detectChallenge(page) {
  let url = "";
  let text = "";
  let markerHits = [];

  try {
    url = page.url();
  } catch { /* page may be mid-navigation */ }

  try {
    text = await page.evaluate(() => (document.body?.innerText || "").substring(0, 12_000));
  } catch { /* ignore — text is best-effort */ }

  try {
    markerHits = await page.evaluate((markers) => {
      // Flatten all selectors across signatures, check in one DOM pass.
      const hits = [];
      for (const sel of markers) {
        try {
          if (document.querySelector(sel)) hits.push(sel);
        } catch { /* invalid selector — skip */ }
      }
      return hits;
    }, CHALLENGE_SIGNATURES.flatMap((s) => s.domMarkers ?? []));
  } catch { /* ignore */ }

  for (const sig of CHALLENGE_SIGNATURES) {
    const urlHit = sig.urlPatterns?.some((p) => p.test(url)) ?? false;
    const textHit = sig.textPatterns?.some((p) => p.test(text)) ?? false;
    const domHit = sig.domMarkers?.some((m) => markerHits.includes(m)) ?? false;

    if (urlHit || textHit || domHit) {
      return { kind: sig.kind, label: sig.label, autoResolvable: sig.autoResolvable };
    }
  }

  // Empty-page heuristic: PerimeterX/imperva and some bot-detection systems
  // serve a near-blank interstitial (often a JS redirect or puzzle iframe)
  // with NO text markers and NO captcha DOM elements. The reliable signal is
  // the combination of TRULY empty visible text AND a title that is just the
  // bare hostname (real pages have descriptive titles).
  //
  // Important: do NOT fire on merely-short content. Chrome's own cert/error
  // interstitials have short text + hostname title but are NOT bot blocks —
  // distinguishing them requires the text to be essentially zero (a PX blank
  // redirect serves no visible text at all). This avoids false-positives on
  // legitimate interstitial pages (badssl cert warnings, etc.).
  const textStripped = text.trim();
  const isTrulyEmpty = textStripped.length === 0;
  if (isTrulyEmpty) {
    let title = "";
    try {
      title = await page.title();
    } catch { /* best-effort */ }
    const bareHost = (() => {
      try { return new URL(url).hostname; } catch { return ""; }
    })();
    // Title equals the bare hostname (e.g. "yelp.com") OR is empty — both are
    // strong indicators of an interstitial block, not real content.
    const titleLooksLikeBlock = !title || title === bareHost || title === bareHost.replace(/^www\./, "");
    if (titleLooksLikeBlock) {
      return { kind: "blocked", label: "Bot-detection empty interstitial", autoResolvable: false };
    }
  }

  return null;
}

/**
 * Wait for an auto-resolvable challenge to clear. Polls detectChallenge() every
 * `pollIntervalMs`; returns as soon as no challenge is detected or the budget
 * expires. Does NOT attempt interactive challenges — those need a solver.
 *
 * @param {import("puppeteer").Page} page
 * @param {{ timeoutMs?: number, pollIntervalMs?: number }} [options]
 * @returns {Promise<{ detected: true, kind: string, label: string, resolved: boolean, waitedMs: number }>}
 */
export async function waitForChallengeResolution(page, options = {}) {
  const timeoutMs = options.timeoutMs ?? 15_000;
  const pollIntervalMs = options.pollIntervalMs ?? 1_500;
  const start = Date.now();

  const initial = await detectChallenge(page);
  if (!initial) {
    return { detected: false, kind: null, label: null, resolved: true, waitedMs: 0 };
  }

  // Interactive / hard challenges cannot self-clear — report immediately so the
  // caller gets a fast, honest failure instead of burning the whole budget.
  if (!initial.autoResolvable) {
    return {
      detected: true,
      kind: initial.kind,
      label: initial.label,
      resolved: false,
      waitedMs: 0,
      interactive: true,
    };
  }

  while (Date.now() - start < timeoutMs) {
    await new Promise((r) => setTimeout(r, pollIntervalMs));
    const recheck = await detectChallenge(page);
    if (!recheck) {
      return {
        detected: true,
        kind: initial.kind,
        label: initial.label,
        resolved: true,
        waitedMs: Date.now() - start,
      };
    }
  }

  return {
    detected: true,
    kind: initial.kind,
    label: initial.label,
    resolved: false,
    waitedMs: Date.now() - start,
    timedOut: true,
  };
}

/**
 * The default challenge-resolution budget for the `auto` strategy. Generous
 * enough for Cloudflare/DataDome (typically 5-10s) but bounded so a stuck
 * challenge fails fast rather than hanging the whole request.
 */
export const DEFAULT_CHALLENGE_TIMEOUT_MS = 15_000;

/**
 * Check whether content is "real" (not a blank/interstitial) by measuring
 * visible text length. Used by the auto strategy to decide whether to wait
 * longer for a SPA/challenge render.
 *
 * @param {import("puppeteer").Page} page
 * @param {{ minLength?: number }} [options]
 * @returns {Promise<boolean>}
 */
export async function hasSubstantiveContent(page, options = {}) {
  const minLength = options.minLength ?? 200;
  try {
    const length = await page.evaluate(
      (min) => (document.body?.innerText || "").trim().length >= min,
      minLength,
    );
    return Boolean(length);
  } catch {
    return false;
  }
}

/**
 * Detect whether the page has an infinite-scroll container (lazy-loaded feed).
 * Heuristic: presence of common sentinel elements OR a "load more" pattern.
 *
 * @param {import("puppeteer").Page} page
 * @returns {Promise<boolean>}
 */
export async function detectInfiniteScroll(page) {
  try {
    return await page.evaluate(() => {
      const sentinels = document.querySelectorAll(
        "[data-pagination], [data-load-more], [data-infinite-scroll], .infinite-scroll, .load-more, #load-more, [role='feed']",
      );
      if (sentinels.length > 0) return true;
      // IntersectionObserver-based lazy loaders are a strong SPA-feed signal.
      const feed = document.querySelector("[class*='feed'], [class*='Feed'], main");
      return Boolean(feed) && document.documentElement.scrollHeight > window.innerHeight * 3;
    });
  } catch {
    return false;
  }
}

File: rust/src/challenge.rs
//! Challenge detection + resolution layer (the explicit "bypass JS challenges"
//! capability). Mirrors `server/core/challenge.js`.
//!
//! Many bot-protection systems (Cloudflare, DataDome, PerimeterX) serve a JS
//! challenge page that auto-resolves after 5-15 seconds WHEN the browser
//! fingerprint passes their automation checks. The stealth stack provides that
//! detection resistance; this module adds the patience: detect the challenge
//! page, wait for self-resolution, and report honestly whether it cleared.

use std::thread;
use std::time::{Duration, Instant};

use headless_chrome::browser::tab::Tab;
use serde::Serialize;

/// Default budget for auto-resolvable challenges (covers Cloudflare/DataDome).
pub const DEFAULT_CHALLENGE_TIMEOUT_MS: u64 = 15_000;

#[derive(Debug, Clone, Serialize)]
pub struct ChallengeInfo {
    pub detected: bool,
    pub kind: Option<String>,
    pub label: Option<String>,
    pub resolved: bool,
    pub waited_ms: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interactive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timed_out: Option<bool>,
}

impl ChallengeInfo {
    /// No challenge detected.
    pub fn none() -> Self {
        Self {
            detected: false,
            kind: None,
            label: None,
            resolved: true,
            waited_ms: 0,
            interactive: None,
            timed_out: None,
        }
    }
}

/// A challenge signature: kind, label, whether it auto-resolves, and the
/// detection signals (URL regex, DOM markers, text regex).
struct Signature {
    kind: &'static str,
    label: &'static str,
    auto_resolvable: bool,
    url_patterns: &'static [&'static str],
    dom_markers: &'static [&'static str],
    text_patterns: &'static [&'static str],
}

/// Compile-once regex cache keyed by pattern string. Built lazily on first
/// detection. Avoids recompiling the same patterns on every page extraction.
/// Keys are owned Strings so the cache can outlive the caller's borrow.
static REGEX_CACHE: std::sync::LazyLock<std::sync::Mutex<std::collections::HashMap<String, regex::Regex>>> =
    std::sync::LazyLock::new(|| std::sync::Mutex::new(std::collections::HashMap::new()));

fn match_pattern(pattern: &str, value: &str) -> bool {
    let mut cache = REGEX_CACHE.lock().expect("regex cache mutex poisoned");
    let re = cache
        .entry(pattern.to_string())
        .or_insert_with(|| regex::Regex::new(pattern).expect("valid challenge regex"));
    re.is_match(value)
}

// Challenge signatures: kind, label, auto-resolvable flag, and detection
// signals. dom_markers are CSS-like selectors checked via evaluate(); url/text
// patterns are regex checked against the page URL and visible text.
const SIGNATURES: &[Signature] = &[
    Signature {
        kind: "cloudflare",
        label: "Cloudflare",
        auto_resolvable: true,
        url_patterns: &["^/cdn-cgi/challenge", "/cdn-cgi/turnstile"],
        dom_markers: &["#cf-challenge", "#challenge-form", "#cf-please-wait", ".cf-browser-verification"],
        text_patterns: &["(?i)just a moment", "(?i)checking your browser", "(?i)cf-challenge"],
    },
    Signature {
        kind: "cloudflare_turnstile",
        label: "Cloudflare Turnstile",
        auto_resolvable: false,
        url_patterns: &[],
        dom_markers: &["iframe[src*='challenges.cloudflare.com']", ".cf-turnstile"],
        text_patterns: &["(?i)verify you are human"],
    },
    Signature {
        kind: "datadome",
        label: "DataDome",
        auto_resolvable: true,
        url_patterns: &["_dd_s\\b", "/_dd\\b"],
        dom_markers: &["iframe[src*='datadome']", "#datadome"],
        text_patterns: &["(?i)datadome", "(?i)please verify you are a person"],
    },
    Signature {
        kind: "perimeterx",
        label: "PerimeterX / HUMAN",
        auto_resolvable: true,
        url_patterns: &["/_pxhl/", "/px\\.js"],
        dom_markers: &["#px-captcha", "iframe[src*='px-captcha']"],
        text_patterns: &["(?i)press.*hold", "(?i)perimeterx"],
    },
    Signature {
        kind: "hcaptcha",
        label: "hCaptcha",
        auto_resolvable: false,
        url_patterns: &[],
        dom_markers: &["iframe[src*='hcaptcha']", ".h-captcha"],
        text_patterns: &["(?i)hcaptcha"],
    },
    Signature {
        kind: "recaptcha",
        label: "reCAPTCHA",
        auto_resolvable: false,
        url_patterns: &[],
        dom_markers: &["iframe[src*='recaptcha']", ".g-recaptcha", "#recaptcha"],
        text_patterns: &["(?i)recaptcha"],
    },
    Signature {
        kind: "blocked",
        label: "Hard block",
        auto_resolvable: false,
        url_patterns: &["/sorry/", "/access[-_]?denied"],
        dom_markers: &[],
        text_patterns: &[
            "(?i)\\baccess denied\\b",
            "(?i)\\b403 forbidden\\b",
            "(?i)\\btemporarily blocked\\b",
            "(?i)\\bsecurity check\\b",
        ],
    },
];

/// Inspect a tab for challenge signatures. Returns the first matching kind or
/// a ChallengeInfo indicating no challenge.
pub fn detect_challenge(tab: &Tab) -> ChallengeInfo {
    let url = tab.get_url();

    // Probe DOM markers and text in one evaluate() round-trip.
    let all_markers: Vec<&'static str> = SIGNATURES
        .iter()
        .flat_map(|s| s.dom_markers.iter().copied())
        .collect();

    let marker_js = format!(
        "(() => {{ const markers = {}; const hits = []; for (const sel of markers) {{ try {{ if (document.querySelector(sel)) hits.push(sel); }} catch (e) {{}} }} return hits; }})()",
        serde_json::to_string(&all_markers).unwrap_or_default()
    );
    let text_js = "(document.body && document.body.innerText || '').substring(0, 12000)";

    let (marker_hits, text): (Vec<String>, String) = match (
        tab.evaluate(&marker_js, true),
        tab.evaluate(text_js, true),
    ) {
        (Ok(m), Ok(t)) => {
            let markers = m
                .value
                .as_ref()
                .and_then(|v| {
                    // evaluate may return the value directly or a JSON string.
                    if let Some(arr) = v.as_array() {
                        serde_json::from_value(serde_json::Value::Array(arr.clone())).ok()
                    } else {
                        v.as_str()
                            .and_then(|s| serde_json::from_str::<Vec<String>>(s).ok())
                    }
                })
                .unwrap_or_default();
            let text_val = t
                .value
                .as_ref()
                .and_then(|v| v.as_str().map(String::from))
                .unwrap_or_default();
            (markers, text_val)
        }
        _ => (Vec::new(), String::new()),
    };

    // Match in signature order; first hit wins.
    for sig in SIGNATURES.iter() {
        let url_hit = sig
            .url_patterns
            .iter()
            .any(|pat| match_pattern(pat, &url));
        let text_hit = sig
            .text_patterns
            .iter()
            .any(|pat| match_pattern(pat, &text));
        let dom_hit = sig
            .dom_markers
            .iter()
            .any(|m| marker_hits.iter().any(|h| h == m));

        if url_hit || text_hit || dom_hit {
            return ChallengeInfo {
                detected: true,
                kind: Some(sig.kind.to_string()),
                label: Some(sig.label.to_string()),
                resolved: false,
                waited_ms: 0,
                interactive: if !sig.auto_resolvable { Some(true) } else { None },
                timed_out: None,
            };
        }
    }

    ChallengeInfo::none()
}

/// Wait for an auto-resolvable challenge to clear. Polls detect_challenge()
/// every `poll_interval_ms`; returns as soon as no challenge is detected or the
/// budget expires. Does NOT attempt interactive challenges.
pub fn wait_for_challenge_resolution(
    tab: &Tab,
    timeout_ms: u64,
    poll_interval_ms: u64,
) -> ChallengeInfo {
    let initial = detect_challenge(tab);
    if !initial.detected {
        return ChallengeInfo::none();
    }

    // Interactive / hard challenges cannot self-clear — report immediately.
    if initial.interactive == Some(true) {
        return initial;
    }

    let start = Instant::now();
    let kind = initial.kind.clone();
    let label = initial.label.clone();

    while (start.elapsed().as_millis() as u64) < timeout_ms {
        thread::sleep(Duration::from_millis(poll_interval_ms));
        let recheck = detect_challenge(tab);
        if !recheck.detected {
            return ChallengeInfo {
                detected: true,
                kind,
                label,
                resolved: true,
                waited_ms: start.elapsed().as_millis() as u64,
                interactive: None,
                timed_out: None,
            };
        }
    }

    ChallengeInfo {
        detected: true,
        kind,
        label,
        resolved: false,
        waited_ms: start.elapsed().as_millis() as u64,
        interactive: None,
        timed_out: Some(true),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn challenge_info_none_serializes_cleanly() {
        let info = ChallengeInfo::none();
        assert!(!info.detected);
        assert!(info.resolved);
    }

    #[test]
    fn signatures_table_is_populated() {
        assert!(!SIGNATURES.is_empty());
        // Cloudflare should be present and auto-resolvable.
        let cf = SIGNATURES.iter().find(|s| s.kind == "cloudflare").unwrap();
        assert!(cf.auto_resolvable);
        // hCaptcha should be present and NOT auto-resolvable.
        let hc = SIGNATURES.iter().find(|s| s.kind == "hcaptcha").unwrap();
        assert!(!hc.auto_resolvable);
    }
}
